File: backend/app/schemas/workflow.py
"""Pydantic schemas for workflows."""

import uuid
from datetime import datetime

from pydantic import BaseModel, Field

from app.models.workflow import WorkflowStatus


class NodePosition(BaseModel):
    x: float
    y: float


class NodeConfig(BaseModel):
    """Permissive dict — each node type has its own config shape."""

    model_config = {"extra": "allow"}


class CanvasNode(BaseModel):
    id: str
    type: str
    label: str
    position: NodePosition
    config: dict = Field(default_factory=dict)


class CanvasEdge(BaseModel):
    source: str
    target: str


class Canvas(BaseModel):
    nodes: list[CanvasNode] = Field(default_factory=list)
    edges: list[CanvasEdge] = Field(default_factory=list)


class WorkflowCreate(BaseModel):
    name: str = Field(..., min_length=1, max_length=255)
    description: str | None = None


class WorkflowUpdate(BaseModel):
    name: str | None = Field(None, min_length=1, max_length=255)
    description: str | None = None
    canvas: Canvas | None = None
    schedule: str | None = None
    tags: list[str] | None = None


class WorkflowResponse(BaseModel):
    id: uuid.UUID
    name: str
    description: str | None
    status: WorkflowStatus
    canvas: Canvas
    schedule: str | None
    tags: list[str]
    created_at: datetime
    updated_at: datetime

    model_config = {"from_attributes": True}


class WorkflowListItem(BaseModel):
    id: uuid.UUID
    name: str
    status: WorkflowStatus
    tags: list[str]
    updated_at: datetime

    model_config = {"from_attributes": True}

File: frontend/src/types/workflow.ts
export type WorkflowStatus = "draft" | "published";

export interface WorkflowListItem {
  id: string;
  name: string;
  status: WorkflowStatus;
  tags: string[];
  updated_at: string;
}

export interface Canvas {
  nodes: CanvasNode[];
  edges: CanvasEdge[];
}

export interface CanvasNode {
  id: string;
  type: string;
  label: string;
  position: { x: number; y: number };
  config: Record<string, unknown>;
}

export interface CanvasEdge {
  source: string;
  target: string;
}

export interface WorkflowResponse {
  id: string;
  name: string;
  description: string | null;
  status: WorkflowStatus;
  canvas: Canvas;
  schedule: string | null;
  tags: string[];
  created_at: string;
  updated_at: string;
}

export interface WorkflowCreate {
  name: string;
  description?: string | null;
}

export interface WorkflowUpdate {
  name?: string | null;
  description?: string | null;
  canvas?: Canvas;
  schedule?: string | null;
  tags?: string[] | null;
}
